src/Microsoft.OData.Core/ODataNotificationWriter.cs

```diff
@@ -324,13 +331,40 @@ protected override void Dispose(bool disposing)
                 if (this.listener != null)
                 {
                     // Tell the listener that the stream is being disposed.
-                    this.listener.StreamDisposed();
+                    if (synchronous)
+                    {
+                        this.listener.StreamDisposed();
+                    }
+                    else
+                    {
+                        this.listener.StreamDisposedAsync().Wait();
+                    }
+
                     this.listener = null;
                 }
             }
 
             this.textWriter.Dispose();

```

Comment:
Why is `this.textWriter.Dispose()` not called in `if (disposing)` block. Also, why isn't it set to `null` after being disposed?

Based on [the guidelines](https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose), I assume the following pattern would have been more appropriate:
```c#
if (disposing) {
   //...
  this.textWriter?.Dispose();
}

// ...
this.textWriter = null;
```

Reply:
@habbes That is a good question but that logic was introduced before by another contributor. I'd be hesitant to change unless there was a good reason to - e.g. a confirmed bug.

Reply:
Fixed